source/common/config/subscription_factory.cc

```diff
@@ -11,11 +11,74 @@
 namespace Envoy {
 namespace Config {
 
+const char UnknownMethod[] = "could_not_lookup_method_due_to_unknown_type_url";
+
+#define TYPE_URL_IS(x)                                                                             \
+  (type_url == Grpc::Common::typeUrl(envoy::api::v2::x().GetDescriptor()->full_name()))
+const Protobuf::MethodDescriptor& deltaGrpcMethod(absl::string_view type_url) {
+  std::string method_name = UnknownMethod;
+  if (TYPE_URL_IS(RouteConfiguration)) {
+    method_name = "envoy.api.v2.RouteDiscoveryService.DeltaRoutes";
+  } else if (TYPE_URL_IS(ScopedRouteConfiguration)) {
+    method_name = "envoy.api.v2.ScopedRoutesDiscoveryService.DeltaScopedRoutes";
+  } else if (TYPE_URL_IS(route::VirtualHost)) {
+    method_name = "envoy.api.v2.VirtualHostDiscoveryService.DeltaVirtualHosts";
+  } else if (TYPE_URL_IS(auth::Secret)) {
+    method_name = "envoy.service.discovery.v2.SecretDiscoveryService.DeltaSecrets";
+  } else if (TYPE_URL_IS(Cluster)) {
+    method_name = "envoy.api.v2.ClusterDiscoveryService.DeltaClusters";
+  } else if (TYPE_URL_IS(ClusterLoadAssignment)) {
+    method_name = "envoy.api.v2.EndpointDiscoveryService.DeltaEndpoints";
+  } else if (TYPE_URL_IS(Listener)) {
+    method_name = "envoy.api.v2.ListenerDiscoveryService.DeltaListeners";
+  }
```

Comment:
Can't we just use the `ConstSingleton` pattern here? I think this is a bit cleaner.

Reply:
Do you mean to replace the string constants, or as some simplification of the conversion? I see V1Converter in there, but that (just adding/removing a prefix) is more programmatic than what's happening here.

Reply:
A few things going on:
1. I think that the subscription factory probably isn't the ideal place for this. I do agree it might be cleaner to avoid the proliferation of strings, but so far the subscription factory has no per-resource special case logic, don't see a reason to start now.
2. The problem you're solving is real, I see the need to make it easier to map from resource type X to the various gRPC, REST, delta gRPC endpoints etc. Can we create a `class TypeUrlEndpoints` that is effectively a tuple of these different resources, and then we just pass this around?

Reply:
I would be completely on board with this being its own class, except that this is the only place where the resource type + transport type = method descriptor formula needs to happen, and I can't immediately think of where else it might be used in the future. So, there is no need for a transport-type-not-yet-determined thing to be passed around. I also don't think such a class's implementation would look any cleaner - it would have to be pretty much these same translator functions. So, I'd rather keep them as is.

I've moved them over to proto_descriptors.cc (which this PR now also updates with the recently added types+services).